File: Sources/Mobi2EpubTransferCore/EpubPreviewParser.swift
import Foundation
import ImageIO

public enum EpubReadingDirection: String, Equatable, Sendable {
    case leftToRight
    case rightToLeft
}

public struct EpubImagePreviewPage: Identifiable, Equatable, Sendable {
    public let id: String
    public let title: String
    public let imageURL: URL
    public let width: Int
    public let height: Int
}

public enum EpubPreviewMode: Equatable, Sendable {
    case imagePages([EpubImagePreviewPage])
    case web(spineURLs: [URL])
}

public struct EpubPreviewBook: Identifiable, Equatable, Sendable {
    public let id = UUID()
    public let title: String
    public let epubURL: URL
    public let extractionDirectory: URL
    public let contentRootDirectory: URL
    public let readingDirection: EpubReadingDirection
    public let mode: EpubPreviewMode
}

public enum EpubPreviewParserError: Error, Equatable, Sendable {
    case unzipFailed(String)
    case missingPackageDocument
    case unreadablePackageDocument
    case missingReadableContent

    public var message: String {
        switch self {
        case .unzipFailed(let log):
            "Could not open EPUB archive. \(log)"
        case .missingPackageDocument:
            "The EPUB package document could not be found."
        case .unreadablePackageDocument:
            "The EPUB package document could not be read."
        case .missingReadableContent:
            "No readable pages were found in this EPUB."
        }
    }
}

public struct EpubPreviewParser {
    private let runner: any ProcessRunning
    private let fileManager: FileManager

    public init(runner: any ProcessRunning = ProcessRunner(), fileManager: FileManager = .default) {
        self.runner = runner
        self.fileManager = fileManager
    }

    public func parse(epubURL: URL, extractionDirectory: URL) async throws -> EpubPreviewBook {
        try fileManager.createDirectory(at: extractionDirectory, withIntermediateDirectories: true)
        let unzipResult = try await runner.run(
            executableURL: URL(fileURLWithPath: "/usr/bin/unzip"),
            arguments: ["-q", epubURL.path, "-d", extractionDirectory.path]
        )
        guard unzipResult.exitCode == 0 else {
            throw EpubPreviewParserError.unzipFailed(unzipResult.output)
        }

        let opfURL = try packageDocumentURL(in: extractionDirectory)
        guard let opf = try? String(contentsOf: opfURL, encoding: .utf8) else {
            throw EpubPreviewParserError.unreadablePackageDocument
        }

        let packageDirectory = opfURL.deletingLastPathComponent()
        let title = metadataTitle(from: opf) ?? epubURL.deletingPathExtension().lastPathComponent
        let manifest = manifestItems(from: opf)
        let spineIDs = spineItemIDs(from: opf)
        let direction = opf.contains(#"page-progression-direction="rtl""#) ? EpubReadingDirection.rightToLeft : .leftToRight

        let spineItems = spineIDs.compactMap { manifest[$0] }
        let imagePages = imagePreviewPages(from: spineItems, packageDirectory: packageDirectory)
        if imagePages.count >= 2, imagePages.count >= max(1, spineItems.count * 4 / 5) {
            return EpubPreviewBook(
                title: title,
                epubURL: epubURL,
                extractionDirectory: extractionDirectory,
                contentRootDirectory: packageDirectory,
                readingDirection: direction,
                mode: .imagePages(imagePages)
            )
        }

        let spineURLs = spineItems.compactMap { item -> URL? in
            guard item.mediaType.contains("xhtml") || item.mediaType.contains("html") else { return nil }
            return EpubPathSecurity.resolve(
                item.href,
                relativeTo: packageDirectory,
                containedIn: extractionDirectory
            )
        }
        if !spineURLs.isEmpty {
            return EpubPreviewBook(
                title: title,
                epubURL: epubURL,
                extractionDirectory: extractionDirectory,
                contentRootDirectory: packageDirectory,
                readingDirection: direction,
                mode: .web(spineURLs: spineURLs)
            )
        }

        throw EpubPreviewParserError.missingReadableContent
    }

    private func packageDocumentURL(in extractionDirectory: URL) throws -> URL {
        let containerURL = extractionDirectory.appendingPathComponent("META-INF/container.xml")
        if
            let containerXML = try? String(contentsOf: containerURL, encoding: .utf8),
            let fullPath = firstAttribute("full-path", inTagMatching: #"<rootfile\b[^>]*>"#, text: containerXML)
        {
            guard let packageURL = EpubPathSecurity.resolve(
                decodeXML(fullPath),
                relativeTo: extractionDirectory,
                containedIn: extractionDirectory
            ) else {
                throw EpubPreviewParserError.missingPackageDocument
            }
            return packageURL
        }

        let candidates = [
            extractionDirectory.appendingPathComponent("content.opf"),
            extractionDirectory.appendingPathComponent("OEBPS/content.opf")
        ]
        if let candidate = candidates.first(where: { fileManager.fileExists(atPath: $0.path) }) {
            return candidate
        }
        throw EpubPreviewParserError.missingPackageDocument
    }

    private func metadataTitle(from opf: String) -> String? {
        firstCapture(pattern: #"<dc:title[^>]*>(.*?)</dc:title>"#, text: opf)
            .map(decodeXML)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .flatMap { $0.isEmpty ? nil : $0 }
    }

    private func manifestItems(from opf: String) -> [String: ManifestItem] {
        tags(matching: #"<item\b[^>]*>"#, in: opf).reduce(into: [:]) { result, tag in
            guard
                let id = attribute("id", in: tag),
                let href = attribute("href", in: tag)
            else {
                return
            }
            result[id] = ManifestItem(
                id: id,
                href: decodeXML(href),
                mediaType: attribute("media-type", in: tag) ?? "",
                properties: attribute("properties", in: tag) ?? ""
            )
        }
    }

    private func spineItemIDs(from opf: String) -> [String] {
        tags(matching: #"<itemref\b[^>]*>"#, in: opf).compactMap { attribute("idref", in: $0) }
    }

    private func imagePreviewPages(from spineItems: [ManifestItem], packageDirectory: URL) -> [EpubImagePreviewPage] {
        spineItems.enumerated().compactMap { index, item in
            guard let itemURL = EpubPathSecurity.resolve(
                item.href,
                relativeTo: packageDirectory,
                containedIn: packageDirectory
            ) else {
                return nil
            }
            let imageURL: URL?
            if item.mediaType.hasPrefix("image/") {
                imageURL = itemURL
            } else if item.mediaType.contains("xhtml") || item.mediaType.contains("html") {
                imageURL = imageURLReferencedByImageOnlyXHTML(at: itemURL)
            } else {
                imageURL = nil
            }

            guard let imageURL, fileManager.fileExists(atPath: imageURL.path) else {
                return nil
            }
            let dimensions = imageDimensions(at: imageURL) ?? (width: 1200, height: 1800)
            let pageNumber = index + 1
            return EpubImagePreviewPage(
                id: String(format: "page-%05d", pageNumber),
                title: "Page \(pageNumber)",
                imageURL: imageURL,
                width: dimensions.width,
                height: dimensions.height
            )
        }
    }

    private func imageURLReferencedByImageOnlyXHTML(at htmlURL: URL) -> URL? {
        guard let data = try? Data(contentsOf: htmlURL) else {
            return nil
        }
        let inspector = XHTMLImagePageInspector()
        let parser = XMLParser(data: data)
        parser.delegate = inspector
        guard parser.parse(), let reference = inspector.imageOnlyReference else {
            return nil
        }
        return EpubPathSecurity.resolve(
            decodeXML(reference),
            relativeTo: htmlURL.deletingLastPathComponent(),
            containedIn: packageRoot(for: htmlURL)
        )
    }

    private func packageRoot(for url: URL) -> URL {
        var current = url.deletingLastPathComponent()
        while current.pathComponents.count > 1 {
            if fileManager.fileExists(atPath: current.appendingPathComponent("content.opf").path) {
                return current
            }
            current.deleteLastPathComponent()
        }
        return url.deletingLastPathComponent()
    }

    private func imageDimensions(at url: URL) -> (width: Int, height: Int)? {
        guard
            let source = CGImageSourceCreateWithURL(url as CFURL, nil),
            let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
            let width = properties[kCGImagePropertyPixelWidth] as? Int,
            let height = properties[kCGImagePropertyPixelHeight] as? Int
        else {
            return nil
        }
        return (width, height)
    }

    private func tags(matching pattern: String, in text: String) -> [String] {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive, .dotMatchesLineSeparators]) else {
            return []
        }
        return regex.matches(in: text, range: NSRange(text.startIndex..<text.endIndex, in: text)).compactMap { match in
            Range(match.range, in: text).map { String(text[$0]) }
        }
    }

    private func firstAttribute(_ name: String, inTagMatching pattern: String, text: String) -> String? {
        tags(matching: pattern, in: text).compactMap { attribute(name, in: $0) }.first
    }

    private func attribute(_ name: String, in tag: String) -> String? {
        firstCapture(pattern: #"\b\#(name)\s*=\s*["']([^"']*)["']"#, text: tag)
    }

    private func firstCapture(pattern: String, text: String) -> String? {
        guard
            let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive, .dotMatchesLineSeparators]),
            let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..<text.endIndex, in: text)),
            match.numberOfRanges > 1,
            let range = Range(match.range(at: 1), in: text)
        else {
            return nil
        }
        return String(text[range])
    }

    private func decodeXML(_ text: String) -> String {
        text
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&apos;", with: "'")
    }
}

private struct ManifestItem: Equatable {
    let id: String
    let href: String
    let mediaType: String
    let properties: String
}

private final class XHTMLImagePageInspector: NSObject, XMLParserDelegate {
    private var ignoredElementDepth = 0
    private var imageReferences: [String] = []
    private var visibleText = ""

    var imageOnlyReference: String? {
        guard
            imageReferences.count == 1,
            visibleText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
            !imageReferences[0].hasPrefix("data:")
        else {
            return nil
        }
        return imageReferences[0]
    }

    func parser(
        _ parser: XMLParser,
        didStartElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?,
        attributes attributeDict: [String: String] = [:]
    ) {
        if ignoredElementDepth > 0 {
            ignoredElementDepth += 1
            return
        }

        let name = localName(qName ?? elementName)
        if ["head", "script", "style"].contains(name) {
            ignoredElementDepth = 1
            return
        }

        guard name == "img" || name == "image" else { return }
        let reference = attributeDict.first { key, _ in
            let attributeName = localName(key)
            return attributeName == "src" || attributeName == "href"
        }?.value
        if let reference {
            imageReferences.append(reference)
        }
    }

    func parser(
        _ parser: XMLParser,
        didEndElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?
    ) {
        if ignoredElementDepth > 0 {
            ignoredElementDepth -= 1
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        guard ignoredElementDepth == 0 else { return }
        visibleText.append(string)
    }

    private func localName(_ name: String) -> String {
        name.split(separator: ":").last.map(String.init)?.lowercased() ?? name.lowercased()
    }
}

File: Windows/src/MobiVerse.Core/EpubPreviewParser.cs
using System.Text.RegularExpressions;
using System.Xml.Linq;

namespace MobiVerse.Core;

public sealed class EpubPreviewParser
{
    private static readonly HashSet<string> ImageExtensions = new(StringComparer.OrdinalIgnoreCase)
    { ".jpg", ".jpeg", ".png", ".gif", ".webp" };

    public EpubPreviewBook Parse(string epubPath, string extractionDirectory)
    {
        Directory.CreateDirectory(extractionDirectory);
        EpubPathSecurity.ExtractSafely(epubPath, extractionDirectory);
        var opfPath = FindPackageDocument(extractionDirectory);
        var document = XDocument.Load(opfPath, LoadOptions.PreserveWhitespace);
        var packageDirectory = Path.GetDirectoryName(opfPath)!;
        var title = document.Descendants().FirstOrDefault(element => element.Name.LocalName == "title")?.Value.Trim();
        if (string.IsNullOrWhiteSpace(title)) title = Path.GetFileNameWithoutExtension(epubPath);

        var manifest = document.Descendants()
            .Where(element => element.Name.LocalName == "item")
            .Where(element => element.Attribute("id") is not null && element.Attribute("href") is not null)
            .GroupBy(element => element.Attribute("id")!.Value, StringComparer.Ordinal)
            .ToDictionary(
                group => group.Key,
                group =>
                {
                    var element = group.First();
                    return new ManifestItem(
                        element.Attribute("href")!.Value,
                        element.Attribute("media-type")?.Value ?? string.Empty,
                        element.Attribute("properties")?.Value ?? string.Empty);
                },
                StringComparer.Ordinal);
        var spineItems = document.Descendants()
            .Where(element => element.Name.LocalName == "itemref")
            .Select(element => element.Attribute("idref")?.Value)
            .Where(value => value is not null && manifest.ContainsKey(value))
            .Select(value => manifest[value!])
            .ToArray();
        var direction = document.Descendants().FirstOrDefault(element => element.Name.LocalName == "spine")?
            .Attribute("page-progression-direction")?.Value.Equals("rtl", StringComparison.OrdinalIgnoreCase) == true
            ? EpubReadingDirection.RightToLeft
            : EpubReadingDirection.LeftToRight;

        var pages = new List<EpubImagePreviewPage>();
        var imageOnlySpineCount = 0;
        foreach (var item in spineItems)
        {
            var itemPath = EpubPathSecurity.Resolve(item.Href, packageDirectory, extractionDirectory);
            if (itemPath is null || !File.Exists(itemPath)) continue;
            IReadOnlyList<string> imagePaths;
            if (item.MediaType.StartsWith("image/", StringComparison.OrdinalIgnoreCase))
            {
                imagePaths = [itemPath];
            }
            else if (item.MediaType.Contains("html", StringComparison.OrdinalIgnoreCase))
            {
                imagePaths = ImageReferencesFromImageOnlyXhtml(itemPath, extractionDirectory);
            }
            else continue;

            if (imagePaths.Count == 0) continue;
            imageOnlySpineCount++;
            foreach (var imagePath in imagePaths.Where(File.Exists).Where(path => ImageExtensions.Contains(Path.GetExtension(path))))
            {
                var dimensions = ReadDimensions(imagePath);
                var number = pages.Count + 1;
                pages.Add(new($"page-{number:00000}", $"Page {number}", imagePath, dimensions.Width, dimensions.Height));
            }
        }

        if (pages.Count >= 2 && imageOnlySpineCount * 5 >= Math.Max(1, spineItems.Length * 4))
            return new(title, epubPath, extractionDirectory, packageDirectory, direction, EpubPreviewKind.ImagePages, pages, []);

        var spinePaths = spineItems
            .Where(item => item.MediaType.Contains("html", StringComparison.OrdinalIgnoreCase))
            .Select(item => EpubPathSecurity.Resolve(item.Href, packageDirectory, extractionDirectory))
            .Where(path => path is not null && File.Exists(path))
            .Cast<string>()
            .ToArray();
        if (spinePaths.Length == 0) throw new InvalidDataException("No readable pages were found in this EPUB.");
        return new(title, epubPath, extractionDirectory, packageDirectory, direction, EpubPreviewKind.Web, [], spinePaths);
    }

    public string? ExtractCover(string epubPath, string extractionDirectory)
    {
        Directory.CreateDirectory(extractionDirectory);
        EpubPathSecurity.ExtractSafely(epubPath, extractionDirectory);
        var opfPath = FindPackageDocument(extractionDirectory);
        var document = XDocument.Load(opfPath);
        var packageDirectory = Path.GetDirectoryName(opfPath)!;
        var items = document.Descendants().Where(element => element.Name.LocalName == "item").ToArray();
        var coverId = document.Descendants().FirstOrDefault(element =>
            element.Name.LocalName == "meta" && element.Attribute("name")?.Value == "cover")?.Attribute("content")?.Value;
        var item = items.FirstOrDefault(element => HasProperty(element, "cover-image"))
            ?? items.FirstOrDefault(element => element.Attribute("id")?.Value == coverId)
            ?? items.FirstOrDefault(element => element.Attribute("id")?.Value.Contains("cover", StringComparison.OrdinalIgnoreCase) == true);
        var href = item?.Attribute("href")?.Value;
        var path = href is null ? null : EpubPathSecurity.Resolve(href, packageDirectory, extractionDirectory);
        return path is not null && File.Exists(path) ? path : null;
    }

    private static string FindPackageDocument(string root)
    {
        var containerPath = Path.Combine(root, "META-INF", "container.xml");
        if (File.Exists(containerPath))
        {
            var container = XDocument.Load(containerPath);
            var fullPath = container.Descendants().FirstOrDefault(element => element.Name.LocalName == "rootfile")?.Attribute("full-path")?.Value;
            var resolved = fullPath is null ? null : EpubPathSecurity.Resolve(fullPath, root, root);
            if (resolved is not null && File.Exists(resolved)) return resolved;
        }
        return new[] { Path.Combine(root, "content.opf"), Path.Combine(root, "OEBPS", "content.opf") }
                   .FirstOrDefault(File.Exists)
               ?? Directory.EnumerateFiles(root, "*.opf", SearchOption.AllDirectories).FirstOrDefault()
               ?? throw new InvalidDataException("The EPUB package document could not be found.");
    }

    private static IReadOnlyList<string> ImageReferencesFromImageOnlyXhtml(string htmlPath, string root)
    {
        try
        {
            var document = XDocument.Load(htmlPath, LoadOptions.PreserveWhitespace);
            var visibleText = string.Concat(document.DescendantNodes().OfType<XText>()
                .Where(node => !node.Ancestors().Any(ancestor =>
                    new[] { "head", "script", "style" }.Contains(ancestor.Name.LocalName, StringComparer.OrdinalIgnoreCase)))
                .Select(node => node.Value)).Trim();
            if (visibleText.Length > 0) return [];
            return document.Descendants()
                .Where(element => element.Name.LocalName.Equals("img", StringComparison.OrdinalIgnoreCase) ||
                                  element.Name.LocalName.Equals("image", StringComparison.OrdinalIgnoreCase))
                .Select(element => element.Attributes().FirstOrDefault(attribute =>
                    attribute.Name.LocalName.Equals("src", StringComparison.OrdinalIgnoreCase) ||
                    attribute.Name.LocalName.Equals("href", StringComparison.OrdinalIgnoreCase))?.Value)
                .Where(value => !string.IsNullOrWhiteSpace(value) && !value.StartsWith("data:", StringComparison.OrdinalIgnoreCase))
                .Select(value => EpubPathSecurity.Resolve(value!, Path.GetDirectoryName(htmlPath)!, root))
                .Where(path => path is not null)
                .Cast<string>()
                .ToArray();
        }
        catch { return []; }
    }

    private static (int Width, int Height) ReadDimensions(string path)
    {
        try { return ImageDimensions.Read(path); }
        catch { return (1200, 1800); }
    }

    private static bool HasProperty(XElement element, string value) =>
        element.Attribute("properties")?.Value.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries)
            .Contains(value, StringComparer.OrdinalIgnoreCase) == true;

    private sealed record ManifestItem(string Href, string MediaType, string Properties);
}
